framework/src/media/media_player.c

```diff
@@ -0,0 +1,320 @@
+/****************************************************************************
+ *
+ * Copyright 2017 Samsung Electronics All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+
+#include <media/media_player.h>
+#include <pthread.h>
+
+#include <tinyara/audio/pcm.h> // For Testing
+
+#define PLAY_LOCK() do { pthread_mutex_lock(&g_pc.mutex_play); } while(false)
+#define PLAY_UNLOCK() do { pthread_mutex_unlock(&g_pc.mutex_play); } while(false)
+
+#define MAX_HEAD_SIZE 1000
+struct media_play_context_s {
+	pthread_t           pth;
+	bool                thread_running;
+	pthread_mutexattr_t mutex_attr_play;
+	pthread_mutex_t     mutex_play;
+	sem_t               sem_pause;
+	struct pcm         *pcmout;
+	unsigned int        buffer_size;
+	int                 fd;
+	int		    download_size;
+	media_format_e      format;
+	play_state_t        state;
+};
+
+static struct media_play_context_s g_pc;
+int threadfunc_media_playing(void *args);
+
+/**
+ * Public Functions
+ */
+play_result_t media_play(int fd, media_format_e format)
+{
+	g_pc.fd = fd;
+	char instruction = 0x00;
+	send(g_pc.fd, &instruction, 1, 0);
+
+	pthread_attr_t attr;
+	struct sched_param sparam;
+
+	PLAY_LOCK();
+
+	if (g_pc.state == PLAY_STOPPING) {
+		pthread_join(g_pc.pth, NULL);
+		g_pc.state = PLAY_AVAILABLE;
+	}
+
+	if (g_pc.state == PLAY_AVAILABLE) {
+
+		g_pc.state = PLAY_RUNNING;
+		g_pc.thread_running = true;
+		if (pthread_attr_init(&attr) != 0) {
+			goto error_thread_init;
+		}
+
+		sparam.sched_priority = CONFIG_MEDIA_PRIORITY;
+		if (pthread_attr_setschedparam(&attr, &sparam) != 0) {
+			goto error_thread_init;
+		}
+
+		if (pthread_attr_setschedpolicy(&attr, SCHED_RR) != 0) {
+			goto error_thread_init;
+		}
+
+		if (pthread_attr_setstacksize(&attr, CONFIG_MEDIA_STACK_SIZE) != 0) {
+			goto error_thread_init;
+		}
+
+		if (pthread_create(&g_pc.pth, &attr, (pthread_startroutine_t)threadfunc_media_playing, NULL) != 0) {

```

Comment:
There is no calling pthread_join or detach. That is a leak.

Reply:
I added the pthread_join to all parts which related to the finish of the thread.